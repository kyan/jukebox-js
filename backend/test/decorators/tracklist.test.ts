import fs from 'fs'
import TransformerTracklist from '../../src/decorators/tracklist'
import ImageCache from '../../src/utils/image-cache'
import Image from '../../src/models/image'
import Mopidy from 'mopidy'
jest.mock('../../src/config/logger')

const firstTrack = {
  _id: 'spotify:track:6IZWJhXyk1Z0rtWNxIi4o7',
  addedBy: [
    {
      user: {
        _id: '123',
        fullname: 'Big Rainbowhead'
      }
    }
  ],
  __v: 0
}
const secondTrack = {
  _id: 'spotify:track:14sOS5L36385FJ3OL8hew4',
  addedBy: [
    {
      user: {
        _id: '456',
        fullname: 'Bigger Rainbowhead'
      }
    }
  ],
  __v: 0
}
const firstImage = {
  _id: 'spotify:track:6IZWJhXyk1Z0rtWNxIi4o7',
  url: 'path/to/image/1'
}
const mockTrackData = [firstTrack, secondTrack]
const mockImageData = [firstImage]

jest.mock('../../src/models/track', () => ({
  findTracks: jest.fn().mockImplementation(() => Promise.resolve(mockTrackData))
}))

describe('TransformerTracklist', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  const payload = JSON.parse(
    fs.readFileSync('./test/__mockData__/tracklist.json', 'utf8')
  )
  describe('when passed a tracklist', () => {
    it('transforms it', () => {
      expect.assertions(1)
      jest.spyOn(ImageCache, 'findAll').mockResolvedValue(mockImageData as Image[])

      return TransformerTracklist(payload).then((transformedPayload) => {
        expect(transformedPayload).toMatchSnapshot()
      })
    })

    it('catches errors', () => {
      expect.assertions(1)
      const arg = 'something broke' as unknown
      return TransformerTracklist(arg as Mopidy.models.Track[]).catch((err) => {
        expect(err.message).toEqual('json.map is not a function')
      })
    })
  })
})
