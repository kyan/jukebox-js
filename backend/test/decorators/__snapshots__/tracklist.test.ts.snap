// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformerTracklist when passed a tracklist transforms it 1`] = `
Array [
  Object {
    "addedBy": Array [
      Object {
        "user": Object {
          "_id": "123",
          "fullname": "Big Rainbowhead",
        },
      },
    ],
    "album": Object {
      "name": "London 0 Hull 4",
      "uri": "spotify:album:13gokJcmO1Dbc9cbHM93jO",
      "year": "1986",
    },
    "artist": Object {
      "name": "The Housemartins",
      "uri": "spotify:artist:77D38RDgCtlYNLpayStftL",
    },
    "image": "path/to/image/1",
    "length": 145000,
    "name": "Happy Hour",
    "uri": "spotify:track:6IZWJhXyk1Z0rtWNxIi4o7",
    "year": "1986",
  },
  Object {
    "addedBy": Array [
      Object {
        "user": Object {
          "_id": "456",
          "fullname": "Bigger Rainbowhead",
        },
      },
    ],
    "album": Object {
      "name": "Happy Now",
      "uri": "spotify:album:6nAjd2MlBY1f1mNu6BsWLO",
      "year": "2018",
    },
    "artist": Object {
      "name": "Kygo",
      "uri": "spotify:artist:23fqKkggKUBHNkbKtXEls4",
    },
    "length": 211000,
    "name": "Happy Now",
    "uri": "spotify:track:14sOS5L36385FJ3OL8hew4",
    "year": "2018",
  },
  Object {
    "album": Object {
      "name": "Eyes Open",
      "uri": "spotify:album:5PYva5C1cdwx2PAsOgZBHN",
      "year": "2006",
    },
    "artist": Object {
      "name": "Snow Patrol",
      "uri": "spotify:artist:3rIZMv9rysU7JkLzEaC5Jp",
    },
    "length": 182000,
    "name": "You Could Be Happy",
    "uri": "spotify:track:7zkLcJktbodeRntovGdXQK",
    "year": "2006",
  },
]
`;
